pkg/controllers/provisioning/v1alpha1/reallocator/controller.go

```diff
@@ -53,31 +55,60 @@ func (c *Controller) Name() string {
 func NewController(kubeClient client.Client, cloudProvider cloudprovider.Factory) *Controller {
 	return &Controller{
 		filter:        &Filter{kubeClient: kubeClient},
+		collector:     &Collector{kubeClient: kubeClient},
 		cloudProvider: cloudProvider,
 	}
 }
 
 // Reconcile executes an allocation control loop for the resource
 func (c *Controller) Reconcile(object controllers.Object) error {
+	provisioner := object.(*v1alpha1.Provisioner)
 	ctx := context.TODO()
-	// 1. Filter all nodes with Karpenter TTL label
-	// TODO: Filter only nodes with the Provisioner label
 
-	// 2. Remove TTL from nodes with TTL label that have pods and if past TTL, cordon/drain node
+	// 1. Filter all nodes with Karpenter labels
+	ttlNodes, err := c.filter.getNodesWithLabels(ctx, []string{
+		provisioning.ProvisionerNamespaceLabelKey,
+		provisioning.ProvisionerNameLabelKey,
+	})
+	if err != nil {
+		return fmt.Errorf("listing ttl provisioner nodes, %w", err)
+	}
+
+	// 2. If node has TTL annotation, and time is after TTL, delete node
+	// TODO: instead of delete - drain then delete
+	for _, node := range ttlNodes.Items {
+		if err := c.collector.ParseTTL(ctx, &node); err != nil {
+			return fmt.Errorf("handling ttl for node, %w", err)
+		}
+	}
 
 	// 3. Filter under-utilized nodes
-	nodes, err := c.filter.GetUnderutilizedNodes(ctx)
+	underutilized, err := c.filter.GetUnderutilizedNodes(ctx, provisioner.Name, provisioner.Namespace)
 	if err != nil {
 		return fmt.Errorf("filtering nodes, %w", err)
 	}
-	if len(nodes) == 0 {
+	if len(underutilized) == 0 {
 		return nil
 	}
-	zap.S().Infof("Found %d underutilized nodes", len(nodes))
 
-	// 4. Cordon each node
+	// TODO: 3.5. Filter underutilized nodes that haven't been cordoned/ttl'd to not spam logs
```

Comment:
3.5?

Reply:
I put 3.5 because it's not really an important step to functionality, but it would log more intelligently.